### PR TITLE
Integrate translation

### PR DESCRIPTION
**Pull request by Zynton**
Related JW PR: https://github.com/odoo-dev/jabberwock/pull/232

### CODE REVIEW

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 1]
```diff
@@ -78,110 +81,92 @@ var Wysiwyg = Widget.extend({
...
-            this.editorHelpers = this.editor.plugins.get(JWEditorLib.DomHelpers);
-
-            // todo: handle megamenu
+        const self = this;
+        const _super = this._super;
+        const elementToParse = document.createElement('div');
+        elementToParse.setAttribute('class', 'd-flex d-column o_editor_center');
+        elementToParse.innerHTML = this.value;
+        elementToParse.style.cssText = 'width: 100%;';
+        $(document.body).addClass('o_connected_user editor_enable');
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: Why have you removed the condition `f (this.options.enableWebsite) {`
If the wysiwyg is instantiated in an "HTML field" or via the website "forum/blog", theses classes should not be added.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 2]
```diff
@@ -349,36 +334,221 @@ var Wysiwyg = Widget.extend({
...
+     */
+    saveTranslationBlocks: async function (reload = true) {
+        const defs = [];
+        this.trigger_up('edition_will_stopped');
+        this.trigger_up('ready_to_save', {defs: defs});
+        await Promise.all(defs);
+
+        const promises = [];
+        await this.editor.execBatch(async () => {
+            await this.editor.execCustomCommand(async () => {
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: I don't see that you are changing the VDocument, nor calling an `execCommand`.
Therfore, why `execBatch` and `execCustomCommand`?

File: addons/website/static/src/js/menu/edit.js [hunk 3]
```diff
@@ -103,7 +151,18 @@ var EditPageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
...
+            plugins: [[this.JWEditorLib.OdooField]],
+            mode: {
+                id: 'edit',
+                rules: [
+                    {
+                        selector: [this.JWEditorLib.DividerNode],
+                        breakable: false,
+                    },
+                ],
+            },
```
Comment (Goaman) [inline addons/website/static/src/js/menu/edit.js]: I think the mode should be encapsulated by the wysiwyg logic. The role of the class wysiwyg is to hide the specific of the editor. But this configuration is specific to our editor.

File: addons/website/static/src/js/menu/edit.js [hunk 4]
```diff
@@ -118,29 +177,89 @@ var EditPageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
...
+                        editable: true,
+                    },
+                    {
+                        selector: [node => {
+                            const attributes = node.modifiers.find(this.JWEditorLib.Attributes);
+                            return attributes && attributes.classList.has('o_not_editable');
+                        }, () => true],
+                        editable: false,
+                    },
+                ],
```
Comment (Goaman) [inline addons/website/static/src/js/menu/edit.js]: Same as above, I would hide this in `wysiwyg.js`

File: addons/website/views/assets.xml [hunk 5]
```diff
@@ -150,6 +150,9 @@
     <script type="text/javascript" src="/website/static/src/js/menu/content.js"/>
     <script type="text/javascript" src="/website/static/src/js/menu/customize.js"/>
     <script type="text/javascript" src="/website/static/src/js/menu/debug_manager.js"/>
+    <!-- todo: move to web_editor.assets_wysiwyg_jabberwock. Currently here because edit
+               depends on it. -->
+    <script type="text/javascript" src="/web_editor/static/lib/jabberwock/jabberwock.js"/>
```
Comment (Goaman) [inline addons/website/views/assets.xml]: It is not needed if you let the logic be encapsulated in `wysiwyg`.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 1 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: It's a mistake on my part, based on a misunderstanding of that condition. I'll put it back.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 2 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: Uh! I guess I used to require a redraw or something. I'll remove it! Good catch!

File: addons/website/static/src/js/menu/edit.js (hunk 3 above)
Comment (Zynton) [reply addons/website/static/src/js/menu/edit.js]: What would you suggest then? Why do we want to hide the params? I need a different configuration for the edit wysiwyg than for the translate wysiwyg. You want to have a `createTranslator` and `createEditor` in `loader.js` then?

File: addons/website/static/src/js/menu/edit.js (hunk 4 above)
Comment (Zynton) [reply addons/website/static/src/js/menu/edit.js]: Same answer as above.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 6]
```diff
@@ -1,10 +1,12 @@
 odoo.define('web_editor.wysiwyg', function (require) {
 'use strict';
+
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:3]: You mean `[REF] web_editor: integrate translations` ?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 7]
```diff
@@ -1,10 +1,12 @@
...
 var Dialog = require('web.Dialog');
 var Widget = require('web.Widget');
 var JWEditorLib = require('web_editor.jabberwock');
 var SnippetsMenu = require('web_editor.snippet.editor').SnippetsMenu;
 var weWidgets = require('wysiwyg.widgets');
+var AttributeTranslateDialog = require('web_editor.wysiwyg.translate_attributes');
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:9]: C'est nouveau ce truc là ?

File: addons/website/static/src/js/menu/edit.js [hunk 8]
```diff
@@ -5,8 +5,45 @@ var core = require('web.core');
...
 var wysiwygLoader = require('web_editor.loader');
 var ajax = require('web.ajax');
+var Dialog = require('web.Dialog');
+var JWEditorLib = require('web_editor.jabberwock');
+var localStorage = require('web.local_storage');
 
-var _t = core._t;
+var localStorageNoDialogKey = 'website_translator_nodialog';
+
+var TranslatorInfoDialog = Dialog.extend({
```
Comment (dmo-odoo) [inline addons/website/static/src/js/menu/edit.js:14]: Ca aussi c'est nouveau ?

File: addons/website/static/src/js/menu/edit.js [hunk 9]
```diff
@@ -5,8 +5,45 @@ var core = require('web.core');
...
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    /**
+     * Called when the "strong" ok is clicked -> adapt localstorage to make sure
+     * the dialog is never displayed again.
+     *
+     * @private
+     */
+    _onStrongOk: function () {
```
Comment (dmo-odoo) [inline addons/website/static/src/js/menu/edit.js:43]: :yes:

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 10]
```diff
@@ -97,12 +97,12 @@ var Wysiwyg = Widget.extend({
             },
             snippetMenuElement: $mainSidebar[0],
             snippetManipulators: $snippetManipulators[0],
-            customCommands: { ...{
+            customCommands: Object.assign({
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:124]: Why ? Genuinely wondering.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 11]
```diff
@@ -430,19 +429,19 @@ var Wysiwyg = Widget.extend({
      * @param {Node} node
      * @returns {Object}
      */
-    _getTranlationObject: function (node) {
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: "La tranlette ça rend sourd."

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 12]
```diff
@@ -40,6 +40,48 @@ var Wysiwyg = Widget.extend({
         this.value = options.value || '';
         this.options = options;
         this.JWEditorLib = JWEditorLib;
+        if (this.options.enableTranslation) {
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:43]: C'est le truc que voulait faire nicolas c'est ça ?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 6 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:3]: Ok :shrug:

File: addons/website/static/src/js/menu/edit.js (hunk 9 above)
Comment (Zynton) [reply addons/website/static/src/js/menu/edit.js:43]: :eyes:

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 7 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:9]: Non, c'était dans `addons/website/static/src/js/editor/wysiwyg_multizone_translate.js` avant.

File: addons/website/static/src/js/menu/edit.js (hunk 8 above)
Comment (Zynton) [reply addons/website/static/src/js/menu/edit.js:14]: Non, c'était dans `editor_menu_translate.js`. D'ailleurs je comprends pas parce que je l'ai supprimé, je ne l'ai plus chez moi, mais je ne vois pas sa suppression dans le diff ici...

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 10 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:124]: The spread syntax for object literals was introduced in ES2018. Odoo is currently stuck in ES2017 so everywhere we use that syntax we should replace it with `Object.assign`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 11 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: :trolleybus:

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 12 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:43]: Oui.

File: addons/website/static/src/js/menu/edit.js (hunk 8 above)
Comment (Zynton) [reply addons/website/static/src/js/menu/edit.js:14]: Ah ben il a été supprimé dans un commit de quelqu'un d'autre apparemment, comme un bourrin...

**Pull request merged**